Addressing issue #29 Helper functions for resource cache

swiftdiaries opened the pull request:
Re-submission of PR, added resource_info.go with helper functions.

k82cn commented:
@swiftdiaries , can you build this PR against master? I'm going to delete 'release-0.1-working' branch.

swiftdiaries commented:
Sure thing. Will do.

jinzhejz commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,71 @@ limitations under the License.
+
+func (r *Resource) Sub(rr *Resource) *Resource {
+	r.MilliCPU -= rr.MilliCPU
+	r.Memory -= rr.Memory
```
suggest check the case `r.MilliCPU < rr.MilliCPU` and `r.Memory < rr.Memory`

swiftdiaries replied:
In the case when both (r.MilliCPU and r.Memory) are less than their counterparts, what should I do with the result? Would it be ok to print out that the resource is not available?

jinzhejz replied:
sure, I think the print log is enough now

k82cn replied:
In MVP, I'd suggest to panic it directly; and then, we need to know how to handle this exception.

For the panic, please create a wrapper func; so we can only log error info in future without go through the whole code.

swiftdiaries replied:
Got it :) Thanks for the suggestions!

jinzhejz commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,71 @@ limitations under the License.
+}
+
+func (r *Resource) Less(rr *Resource) bool {
+	return r.MilliCPU < rr.MilliCPU && r.Memory < rr.Memory
```
this may return `false` for `A.Less(B)` and `B.Less(A)`, such as `A(MilliCPU:2, Memory:1)`, `B(MilliCPU:2, Memory:1)`

k82cn replied:
I think that's what we want. When allocating resource to Pod, we need to know whether we can launch Pod with resources.

jinzhejz replied:
get the point :)

jinzhejz commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,71 @@ limitations under the License.
+
+func (r *Resource) LessEqual(rr *Resource) bool {
+	return (r.MilliCPU < rr.MilliCPU || math.Abs(rr.MilliCPU-r.MilliCPU) < 0.01) &&
+		(r.Memory < rr.Memory || math.Abs(rr.Memory-r.Memory) < 1)
```
suggest remove `Less()` and `LessEqual()` temporarily, or compare `MilliCPU` and `Memory` independently

swiftdiaries replied:
Meaning wherever we do need to compare the two Resources, we do it manually each time?

jinzhejz replied:
as @k82cn mentioned in above comments, I misunderstand the usage of `Less()` and `LessEqual()`, we can keep the current behavior

swiftdiaries replied:
Ok :) 👍

k82cn commented:
xref #29

k82cn commented:
@swiftdiaries , are you still working on this PR?

swiftdiaries commented:
I have edited according to suggested comments. Is there anything else I
need to add?

On Thu, Sep 28, 2017 at 3:48 AM Klaus Ma <notifications@github.com> wrote:

> @swiftdiaries <https://github.com/swiftdiaries> , are you still working
> on this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#issuecomment-332756565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV9Rn9XOi5or1uylT_230mFGvrEkqOTCks5sm08wgaJpZM4PaeBW>
> .
>

k82cn commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,81 @@ limitations under the License.
+var minMilliCPU float64 = 10
+var minMemory float64 = 10 * 1024 * 1024
+
+//Changed NewResource to account for single resource requests
```
What's purpose of this comments?

k82cn commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,81 @@ limitations under the License.
+	return r
+}
+
+//Added panic to show when Resource is insufficient
```
a TODO?

k82cn commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,81 @@ limitations under the License.
+
+//Added panic to show when Resource is insufficient
+
+func (r *Resource) Sub(rr *Resource) *Resource {
```
add comments for the behaviours of Sub.

k82cn commented on pkg/schedulercache/resource_info.go:
```diff
@@ -16,5 +16,81 @@ limitations under the License.
+		r.Memory -= rr.Memory
+		return r
+	}
+	panic("Resource is not sufficient to do operation: Sub()")
```
Add a wrapper func for it; we'll remove panic when it's stable.

swiftdiaries commented:
Oh those were addressing review comments.

On Thu, Sep 28, 2017 at 9:58 AM Klaus Ma <notifications@github.com> wrote:

> *@k82cn* commented on this pull request.
> ------------------------------
>
> In pkg/schedulercache/resource_info.go
> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#discussion_r141625386>
> :
>
> > +		Memory:   0,
> +	}
> +}
> +
> +func (r *Resource) Clone() *Resource {
> +	clone := &Resource{
> +		MilliCPU: r.MilliCPU,
> +		Memory:   r.Memory,
> +	}
> +	return clone
> +}
> +
> +var minMilliCPU float64 = 10
> +var minMemory float64 = 10 * 1024 * 1024
> +
> +//Changed NewResource to account for single resource requests
>
> What's purpose of this comments?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#pullrequestreview-65868182>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV9Rn7J3dipFM76S6BoYGFnUgmpMPry5ks5sm6YhgaJpZM4PaeBW>
> .
>

swiftdiaries commented:
Sure thing. I'll add them and update. Thanks for the review.

On Thu, Sep 28, 2017 at 10:00 AM Adhita D Selvaraj <adhita94@gmail.com>
wrote:

> Oh those were addressing review comments.
>
> On Thu, Sep 28, 2017 at 9:58 AM Klaus Ma <notifications@github.com> wrote:
>
>> *@k82cn* commented on this pull request.
>> ------------------------------
>>
>> In pkg/schedulercache/resource_info.go
>> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#discussion_r141625386>
>> :
>>
>> > +		Memory:   0,
>> +	}
>> +}
>> +
>> +func (r *Resource) Clone() *Resource {
>> +	clone := &Resource{
>> +		MilliCPU: r.MilliCPU,
>> +		Memory:   r.Memory,
>> +	}
>> +	return clone
>> +}
>> +
>> +var minMilliCPU float64 = 10
>> +var minMemory float64 = 10 * 1024 * 1024
>> +
>> +//Changed NewResource to account for single resource requests
>>
>> What's purpose of this comments?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#pullrequestreview-65868182>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AV9Rn7J3dipFM76S6BoYGFnUgmpMPry5ks5sm6YhgaJpZM4PaeBW>
>> .
>>
>

k82cn commented:
And please
1. remove the dependency on `vendor/github.com/ugorji/*`
2. merge those 4 commits into one  (here's the command I use: `git reset`, `git commit`, `git push --force`)

swiftdiaries commented:
Will do that. Thanks for the pointers.

On Thu, Sep 28, 2017 at 10:06 AM Klaus Ma <notifications@github.com> wrote:

> And please
>
>    1. remove the dependency on vendor/github.com/ugorji/*
>    2. merge those 4 commits into one (here's the command I use: git reset,
>    git commit, git push --force)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#issuecomment-332847101>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV9Rn_CLkn1r1pYARptxx37ftAUjRyJeks5sm6fbgaJpZM4PaeBW>
> .
>

k82cn commented:
1. `Godeps/Godeps.json` is conflict
1. why introduce `github.com/ugorji/`?
1. Please follow golang's comment style :).

k82cn commented:
@swiftdiaries , I'd like to get this PR merged after removing the diff about vendor; would you help on that.

swiftdiaries commented:
Yeah, on it. Will get it done.

On Mon, Oct 16, 2017 at 10:32 PM Klaus Ma <notifications@github.com> wrote:

> @swiftdiaries <https://github.com/swiftdiaries> , I'd like to get this PR
> merged after removing the diff about vendor; would you help on that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/kube-arbitrator/pull/38#issuecomment-337099504>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV9RnzCiSD0DLcikQy45968k3VKLsna_ks5stBG7gaJpZM4PaeBW>
> .
>

swiftdiaries commented:
Closing and creating a new one.
